File: api/js/src/validator/rules/empty-row.ts
import { createDiagnostic } from "../diagnostics.js";
import type { ValidationRule } from "../types.js";
import { RULE_IDS } from "./constants.js";

export const emptyRowRule: ValidationRule = {
  id: RULE_IDS.EMPTY_ROW,
  description: "Detects rows without buttons after layout",
  defaultSeverity: "warning",
  run(ctx) {
    const diagnostics = [];

    for (let i = 0; i < ctx.normalized.rows.length; i++) {
      const row = ctx.normalized.rows[i];
      if (row && row.length === 0) {
        diagnostics.push(
          createDiagnostic(
            RULE_IDS.EMPTY_ROW,
            `Row ${i + 1} has no buttons`,
            "warning",
            { row: i },
          ),
        );
      }
    }

    let pendingBreak = false;
    for (const raw of ctx.normalized.rawButtons) {
      if (raw.__newRow) {
        if (pendingBreak) {
          diagnostics.push(
            createDiagnostic(
              RULE_IDS.EMPTY_ROW,
              "Consecutive newRow() markers can produce empty rows",
              "warning",
            ),
          );
        }
        pendingBreak = true;
        continue;
      }
      pendingBreak = false;
    }

    return diagnostics;
  },
};

File: api/python/telegram_inline_keyboard_builder/validator/rules/empty_row.py
from __future__ import annotations

from ..diagnostics import create_diagnostic
from ..types import Diagnostic, RuleContext, ValidationRule
from .constants import RULE_IDS


def _run(ctx: RuleContext) -> list[Diagnostic]:
    diagnostics: list[Diagnostic] = []

    for i, row in enumerate(ctx.normalized.rows):
        if len(row) == 0:
            diagnostics.append(
                create_diagnostic(
                    RULE_IDS.EMPTY_ROW,
                    f"Row {i + 1} has no buttons",
                    "warning",
                    {"row": i},
                )
            )

    pending_break = False
    for raw in ctx.normalized.raw_buttons:
        if raw.get("__newRow"):
            if pending_break:
                diagnostics.append(
                    create_diagnostic(
                        RULE_IDS.EMPTY_ROW,
                        "Consecutive newRow() markers can produce empty rows",
                        "warning",
                    )
                )
            pending_break = True
            continue
        pending_break = False

    return diagnostics


empty_row_rule = ValidationRule(
    id=RULE_IDS.EMPTY_ROW,
    description="Detects rows without buttons after layout",
    default_severity="warning",
    run=_run,
)
